cdc/cdc.cc

```diff
@@ -0,0 +1,197 @@
+        for (const auto& endpoint : dc.second) {
+            f(endpoint);
+        }
+    }
+}
+
+template <typename Func>
+static void for_each_shard(const gms::inet_address& endpoint, Func&& f) {
+    auto& snitch_ptr = locator::i_endpoint_snitch::get_local_snitch_ptr();
+    const unsigned int shard_count = snitch_ptr->get_shard_count(endpoint);
+    for (unsigned int shard_id = 0; shard_id < shard_count; ++shard_id) {
+        f(shard_id, shard_count);
+    }
+}
+
+static future<>
+populate_desc(service::storage_proxy& proxy, const schema& s) {
+    auto& db = proxy.get_db().local();
+    auto desc_schema =
+        db.find_schema(s.ks_name(), desc_name(s.cf_name()));
+    auto log_schema =
+        db.find_schema(s.ks_name(), log_name(s.cf_name()));
+    auto& token_metadata =
+        service::get_local_storage_service().get_token_metadata();
+    auto& replication_strategy =
+        db.find_keyspace(s.ks_name()).get_replication_strategy();
+    // The assumption here is that all nodes in the cluster are using the
+    // same murmur3_partitioner_ignore_msb_bits. If it's not true, we will
+    // have to gossip it around together with shard_count.
+    auto ignore_msb_bits =
+        db.get_config().murmur3_partitioner_ignore_msb_bits();
+
+    auto belongs_to = [&](const gms::inet_address& endpoint,
+                          const unsigned int shard_id,
+                          const int shard_count,
+                          const utils::UUID& stream_id) {
+        const auto log_pk = partition_key::from_singular(*log_schema,
+                                                         data_value(stream_id));
+        const auto token =
+            dht::global_partitioner().decorate_key(*log_schema, log_pk).token();
+        auto&& eps =
+            replication_strategy.calculate_natural_endpoints(token, token_metadata);
+        if (eps.empty() || eps[0] != endpoint) {
+            return false;
+        }
+        const auto owning_shard_id = dht::murmur3_partitioner::shard_of(
+                token, shard_count, ignore_msb_bits);
+        return owning_shard_id == shard_id;
+    };
+
+    std::vector<mutation> mutations;
+    const auto ts = api::new_timestamp();
+    const auto ck = clustering_key::from_single_value(
+            *desc_schema, timestamp_type->decompose(ts));
+    auto cdef = desc_schema->get_column_definition(to_bytes("stream_id"));
+
+    for_each_endpoint([&] (const gms::inet_address& endpoint) {
+        const auto decomposed_ip = inet_addr_type->decompose(endpoint.addr());
+
+        for_each_shard(endpoint, [&] (unsigned int shard_id, int shard_count) {
+            const auto pk = partition_key::from_exploded(
+                    *desc_schema, { decomposed_ip, int32_type->decompose(static_cast<int>(shard_id)) });
+            mutations.emplace_back(desc_schema, pk);
+
+            auto stream_id = utils::make_random_uuid();
+            while (!belongs_to(endpoint, shard_id, shard_count, stream_id)) {
```

Comment:
After brainstorming with @kbr- we came to the conclusion that this can be done better by generating stream_ids for all the endpoints and shards at the same time instead of generating it one by one. Will improve and upload soon.

Reply:
You should have a thread::maybe_yield() here (and run this in a thread), or @glommer will be sad

Reply:
Perhaps each shard should be responsible for populating its own entry in the table. This works better with replica-side logging: when a replica is notified it becomes the owner of a token range, it initializes the stream id for it.

Reply:
Kamil will change this in exactly that way in his patchset. Each shard will be responsible for its streamid.